clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -537,6 +536,7 @@ public void handle(SyncGroupResponse syncResponse,
             if (error == Errors.NONE) {
                 sensors.syncLatency.record(response.requestLatencyMs());
                 future.complete(syncResponse.memberAssignment());
+                AbstractCoordinator.this.rejoinNeeded = false;
```

Comment:
On a second thought... `future.complete(syncResponse.memberAssignment())` above will trigger `joinFuture.addListener`'s `onSuccess`, which will enable the heartbeat thread right away, and hence there is a (very small) race condition.

I think it is safer to just move the the above line inside `onSuccess` (line 395) to set it before enabling heart beat thread, and we would not need `AbstractCoordinator.this` prefix also.